### ISSUE TITLE
model inference -- fps

### ISSUE THREAD

**Issue by uuu686** [stefan-ainetter/grasp_det_seg_cnn#23]
hello, I am confused in some question:
I trained the model with my data, it contained two objects. I test the model on GTX 1050ti, the model spend about 420 ms to inference every frame. That's mean 2-3 fps. Is the result normal? 
I test every module cost time, the resnet101 cost about 400 ms, the RPN cost about 25 ms, the ROI cost about 25 ms,the segmentation cost about 92 ms. The pytorch parallel processing can speed up so the whole model cost about 400 ms-500 ms.
I found you test the model on RTX 2080ti and it can achieve 48 fps.
Is this a problem with my environment, or a graphics card performance problem？
Thanks！

**Comment by stefan-ainetter** [comment]
Hi @uuu686,
can you share the config file that you use for your experiments? For sure the hardware makes a difference, but it should not be as big as in your example.

**Comment by uuu686** [comment]
[general]
#Number of epochs between validations
val_interval = 25
#Number of steps before outputting a log entry
log_interval = 10
cudnn_benchmark = no
num_classes = 18
num_stuff = 0
num_things = 18
#0 - 31  todo
num_semantic = 3


[body]
#Architecture for the body todo
body = resnet101
#Path to pre-trained ImageNet weights todo
weights = /mnt/work_dir/grasp_det_seg_cnn/weights_pretrained/resnet101
#Normalization mode:
#-- bn: in-place batch norm everywhere
#-- syncbn: synchronized in-place batch norm everywhere
#-- syncbn+bn: synchronized in-place batch norm in the static part of the network, in-place batch norm everywhere else
#-- gn: group norm everywhere
#-- syncbn+gn: synchronized in-place batch norm in the static part of the network, group norm everywhere else
#-- off: do not normalize activations (scale and bias are kept)
normalization_mode = syncbn
#Activation: 'leaky_relu' or 'elu'  default is leaky_relu  todo
#activation = elu
activation = leaky_relu
activation_slope = 0.01
#Group norm parameters
gn_groups = 16
#Additional parameters for the body
body_params = {}
#Number of frozen modules: in [1, 5]  # default is 2  todo
num_frozen = 0
#Whether to freeze BN modules todo
bn_frozen = no
#Meta-info
out_channels = {"mod1": 64, "mod2": 256, "mod3": 512, "mod4": 1024, "mod5": 2048}
out_strides = {"mod1": 4, "mod2": 4, "mod3": 8, "mod4": 16, "mod5": 32}

[fpn]
out_channels = 256
extra_scales = 0
interpolation = nearest
#Input settings
inputs = ["mod2", "mod3", "mod4", "mod5"]
#Meta-info
out_strides = (4, 8, 16, 32)

[rpn]
hidden_channels = 256
stride = 1
#Anchor settings
anchor_ratios = (1., 0.1, 0.4, 0.7, 1.2)
anchor_scale = 2
#Proposal settings
nms_threshold = 0.7
num_pre_nms_train = 12000
num_post_nms_train = 2000
num_pre_nms_val = 6000
num_post_nms_val = 300
min_size = 16
#Anchor matcher settings
num_samples = 256
pos_ratio = .5
pos_threshold = .7
neg_threshold = .3
void_threshold = 0.7
#FPN-specific settings
fpn_min_level = 0
fpn_levels = 3
#Loss settings
sigma = 3.

[roi]
roi_size = (14, 14)
#Matcher settings
num_samples = 128
pos_ratio = .25
pos_threshold = .5
neg_threshold_hi = .5
neg_threshold_lo = 0.
void_threshold = 0.7
void_is_background = no
#Prediction generator settings
nms_threshold = 0.3
score_threshold = 0.05
max_predictions = 100
#FPN-specific settings
fpn_min_level = 0
fpn_levels = 4
fpn_canonical_scale = 224
fpn_canonical_level = 2
#Loss settings
sigma = 1.
bbx_reg_weights = (10., 10., 5., 5.)

[sem]
fpn_min_level = 0
fpn_levels = 4
pooling_size = (64, 64)
#Loss settings
ohem = .25

[optimizer]
#todo
#lr = 0.03
lr = 0.005
#try decay = 0.001
weight_decay = 0.0001
weight_decay_norm = yes
momentum = 0.9
nesterov = yes
#obj, bbx, roi_cls, roi_bbx, sem
#loss_weights = (1., 1., 1., 1.,.75)
loss_weights = (.5, .5, 1., .5,.75)

[scheduler]
#800 todo
epochs = 810
#Scheduler type: 'linear', 'step', 'poly' or 'multistep'
type = poly
#When to update the learning rate: 'batch', 'epoch'
update_mode = batch
#Additional parameters for the scheduler
#-- linear
#from: initial lr multiplier
#to: final lr multiplier
#-- step
#step_size: number of steps between lr decreases
#gamma: multiplicative factor
#-- poly
#gamma: exponent of the polynomial
#-- multistep
#milestones: step indicies where the lr decreases will be triggered
params = {"gamma": 0.9}
burn_in_steps = 500
burn_in_start = 0.333

[dataloader]
#数据流 参数
#Absolute path to the project data
root_path = /mnt/work_dir/my_data/OCID_grasp
#Image size parameters  [480, 640]
shortest_size = 480
longest_max_size = 640
#Batch size todo
train_batch_size = 2
val_batch_size = 1
#Augmentation parameters  the parameters is calculate from ImageNet
rgb_mean = (0.485, 0.456, 0.406)
rgb_std = (0.229, 0.224, 0.225)
random_flip = no
random_scale = None
rotate_and_scale = True
#Number of worker threads
num_workers = 6
#Subsets todo
#train_set = mydata273_val
train_set = mydata273_train
val_set = mydata273_val
test_set = validation_0

**Comment by uuu686** [comment]
My configuration file is as shown above, is there anything wrong with it ?

**Comment by uuu686** [comment]
I just tested the model on a 2080Ti and each frame took about 80-90ms. The paper mentioned it takes about 20 ms (about 50fps).
I used the configuration file above.
Did I do something wrong ?

**Comment by uuu686** [comment]
hello, can you point out the problems above

**Comment by stefan-ainetter** [comment]
Hi @uuu686,
one way to  increase the inference speed is to reduce the size of your input image by downsampling or cropping. 

According to the 50fps in the paper: To be able to calculate the maximum frames per second that can be processed by the network, we set the batch size to > 1 . This way, it is possible to exploit the full computational capabilities of the network. However, if you use a camera system which delivers only one frame at a time (which means batch size = 1), I am not sure if you can reach the 20ms.

**Comment by uuu686** [comment]
Thank you. I think I know what you mean.

**Comment by uuu686** [comment]
hello @stefan-ainetter ,
 if I want to reduce the cost time of model inference, what are the ways I can do it?
I use the restnet101 as the backbone, and the size of  input image is 480*640. 
I try to replace the ABN module as torch.batchNorm, and the time of inference is decline.
In addition to the above way, which modules  I can modify to recude the time of inference?
